### ISSUE TITLE
Emitted assembly for function calls is wrong

### ISSUE THREAD

**Issue by bwesterb** [llvm-dcpu16/llvm-dcpu16#163]
When I compile

``` c

int main3(void)
{
        return 42;
}

int main2(void)
{
        return main3();
}

int main(void)
{
        return main2();
}
```

The emitted assembly is

``` dasm16
        ; .file "main.c"                                                         
        .text                                                                    
        .globl  main3                                                            
        ; .align        1                                                        
:main3                                                                           
        SET     A, 0x2a                                                          
        SET     PC, POP                                                          

        .globl  main2                                                            
        ; .align        1                                                        
:main2                                                                           
        SET     PUSH, J                                                          
        SET     J, SP                                                            
        SUB     SP, 0xfffe                                                       
        JSR     main3                                                            
        ADD     SP, 0xfffe                                                       
        SET     J, POP                                                           
        SET     PC, POP                                                          

        .globl  main                                                             
        ; .align        1                                                        
:main                                                                            
        SET     PUSH, J                                                          
        SET     J, SP                                                            
        SET     [0x1+J], 0x0                                                     
        JSR     main2                                                            
        SET     J, POP                                                           
        SET     PC, POP
```

which is wrong for several reasons.  For instance:  `SET[0x1+J],0x0` will prevent `main()` from returning to the caller.

**Comment by ghost** [comment]
This is also a duplicate of #153 . He calculcates wrong offsets for the stack and framepointer and writes at wrong positions. BTW: I think you used -O1 or -O2, since the `SUB     SP, 0xfffe` only occures in those cases. This is all symptom of our change to a wide char (16 bit) and the hard coded parts in LLVM, which thinks that char is only 8 bit wide.

**Comment by bwesterb** [comment]
I did not set an -O flag.  If I do, LLVM does not generate function calls, but simply returns 0x2a for all three functions.

**Comment by CJKay** [comment]
Using `clang -fomit-frame-pointer` appears to fix this issue temporarily (albeit obviously removing the base pointer).
It generates...

```
:main
    SUB SP, 0x2
    SET PICK 0x1, 0x0
    JSR main2
    ADD SP, 0x2
    SET PC, POP

:main2
    JSR main3
    SET PC, POP

:main3
    SET A, 0x2a
    SET PC, POP
```

... which runs fine on all the emulators I've used.

**Comment by Blei** [comment]
Fixed by #167
